src/V3PerPackage/V3PerBatchProcessor.cs

```diff
@@ -0,0 +1,290 @@
+        private readonly StringLocker _stringLocker;
+        private readonly ILogger<V3PerBatchProcessor> _logger;
+
+        public V3PerBatchProcessor(
+            HttpMessageHandler httpMessageHandler,
+            ILoggerFactory loggerFactory,
+            TelemetryClient telemetryClient,
+            StringLocker stringLocker,
+            ILogger<V3PerBatchProcessor> logger)
+        {
+            _httpMessageHandler = httpMessageHandler;
+            _loggerFactory = loggerFactory;
+            _telemetryClient = telemetryClient;
+            _stringLocker = stringLocker;
+            _logger = logger;
+        }
+
+        public async Task<bool> ProcessAsync(PerBatchContext context, IReadOnlyList<PerPackageContext> packageContexts)
+        {
+            var catalogIndexUri = await ExecuteFeedToCatalogAsync(context, packageContexts);
+
+            await ExecuteCatalog2DnxAsync(context, packageContexts, catalogIndexUri);
+
+            await ExecuteCatalog2RegistrationAsync(context, catalogIndexUri);
+
+            await ExecuteCatalog2LuceneAsync(context, catalogIndexUri);
+
+            return true;
+        }
+
+        public async Task CleanUpAsync(PerBatchContext context, IReadOnlyList<PerPackageContext> packageContexts)
+        {
+            var blobClient = BlobStorageUtilities.GetBlobClient(context.Global);
+
+            // Delete catalog2lucene artifacts.
+            await CleanUpUtilities.DeleteContainer(blobClient, context.LuceneContainerName, _logger);
+
+            var luceneCachePath = Path.Combine(
+                CleanUpUtilities.GetLuceneCacheDirectory(),
+                context.LuceneContainerName);
+            Directory.Delete(luceneCachePath, recursive: true);
+
+            // Delete catalog2registration artifacts.
+            await CleanUpUtilities.DeleteBlobsWithPrefix(
+                blobClient,
+                context.Global.RegistrationContainerName,
+                context.Worker.Name,
+                _logger);
+
+            // Delete catalog2dnx artifacts.
+            foreach (var packageContext in packageContexts)
+            {
+                await CleanUpUtilities.DeleteBlobsWithPrefix(
+                    blobClient,
+                    context.Global.FlatContainerContainerName,
+                    $"{context.Process.FlatContainerStoragePath}/{packageContext.PackageId.ToLowerInvariant()}/{packageContext.PackageVersion.ToLowerInvariant()}",
+                    _logger);
+            }
+
+            // Delete feed2catalog artifacts.
+            await CleanUpUtilities.DeleteBlobsWithPrefix(
+                blobClient,
+                context.Global.CatalogContainerName,
+                context.Worker.Name,
+                _logger);
+        }
+
+        private async Task<Uri> ExecuteFeedToCatalogAsync(PerBatchContext context, IReadOnlyList<PerPackageContext> packageContexts)
+        {
+            var serviceProvider = GetServiceProvider(
```

Comment:
>serviceProvider [](start = 16, length = 15)

Can you reuse a method's `serviceProvider` across method calls? I get the feeling that regenerating providers like this is not cheap.